src/pydicom/dataset.py

```diff
@@ -3673,3 +3728,82 @@ def _tsyntax_encoding(self) -> tuple[bool, bool] | tuple[None, None]:
     "(?P<start>(^[A-Za-z])((?=.+?[A-Z])[A-Za-z0-9]+)|(^[A-Z])([A-Za-z0-9]+))"
     "(?P<last>[A-Za-z0-9][^_]$)"  # Last character is alphanumeric
 )
+
+
+def _path_to(target, node) -> str | None:
+    """Return a pseudo-code path to a particular DataElement, Sequence, or value
+
+    This is used by Dataset's context manager to report the specific data element
+    where an error occurred.
+
+    Params
+    ------
+    target: Any
+        Usually a DataElement, Sequence, or Sequence item.  Can be a value,
+        in which case the path to the first DataElement found (depth first search)
+        with that value is returned.
+    node: Any
+        The object currently being searched (function is called recursively).
+        On the first call to the function it will typically be a Dataset
+
+    Returns
+    -------
+    str | None:
+        The path to the target object from the node.
+        During recursion, returns ``None`` if a leaf node is reached without finding target.
+
+    Examples
+    --------
+    >> path_to(data_element)
+    'FileDataset(filename='rtplan.dcm').BeamSequence[0].BeamName'
+
+    """
+    # Recurse down DICOM's nested structures, depth first,
+    # until the target object or value is found.
+    # As the recursion is unwound, the node or index at each level
+    # is prepended to the ongoing "path"
+
+    if not target:
+        return None
+
+    ns = SimpleNamespace(target=target)  # make dotted lookup to avoid name binding
+    match node:
+        case _ if node is target:  # finished if match on identity
+            return ""
+        case Dataset():
+            # recurse into Dataset and file_meta data elements
+            if hasattr(node, "file_meta"):
+                items = chain(node.items(), node.file_meta.items())
+            else:
+                items = node.items()
+            for tag, dataelem in items:
+                if (path := _path_to(target, dataelem)) is not None:
+                    break
+            else:  # for-else = "no break", i.e. target not found
```

Comment:
it's is a testimony to how not intuitive this is, if you have to explain a standard language feature 😁